KAFKA-2475: Make Copycat only have a Converter class instead of Serializer, Deserializer, and Converter.

ewencp opened the pull request:
The Converter class now translates directly between byte[] and Copycat's data
API instead of requiring an intermediate runtime type like Avro's GenericRecord
or Jackson's JsonNode.

ewencp committed:
KAFKA-2475: Make Copycat only have a Converter class instead of Serializer, Deserializer, and Converter.

The Converter class now translates directly between byte[] and Copycat's data
API instead of requiring an intermediate runtime type like Avro's GenericRecord
or Jackson's JsonNode.

gwenshap commented on copycat/api/src/main/java/org/apache/kafka/copycat/data/CopycatSchema.java:
```diff
@@ -161,11 +175,19 @@ public static void validateValue(Schema schema, Object value) {
         // Special case for bytes. byte[] causes problems because it doesn't handle equals()/hashCode() like we want
         // objects to, so we support both byte[] and ByteBuffer. Using plain byte[] can cause those methods to fail, so
         // ByteBuffers are recommended
```
the doc has to go together with the special case?

ewencp replied:
Ah, really a leftover comment that should have been removed, but I've adjusted and moved it up by the list of BYTE classes in SCHEMA_TYPE_CLASSES.

gwenshap commented on copycat/api/src/main/java/org/apache/kafka/copycat/data/SchemaBuilder.java:
```diff
@@ -346,7 +347,7 @@ public Schema valueSchema() {
      */
     public Schema build() {
-        return new CopycatSchema(type, isOptional(), defaultValue, name, version, doc, fields, keySchema, valueSchema);
+        return new CopycatSchema(type, isOptional(), defaultValue, name, version, doc, fields == null ? null : Collections.unmodifiableList(fields), keySchema, valueSchema);
```
that is how we do schema-less? a schema without fields? or is this for a different use-case?

ewencp replied:
Most schemas don't have fields -- all the primitive types, arrays, maps. i.e. everything but structs. In all those cases the list of fields is null (could just as easily be an empty list, especially since I think I have all code paths in Schema that could access it incorrectly throw exceptions if the schema type isn't a struct, but having an empty list doesn't benefit us at all).

This just got added because it's a guard against one of the few places where Schemas were still technically mutable. We could also move this into the CopycatSchema constructor instead, now that I think about it.

gwenshap replied:
got it. +1 for enforcing immutable.

gwenshap commented:
For the file connector, and as a simple example, do we want to add a StringConverter?

ewencp commented:
Debatable, I think, for a few reasons.
1. I think it's bad to use raw types anyway since it gives you few options for updating the schema if you ever need to.
2. Would StringConverter support anything besides String? We could have it convert the primitive types to Strings, which wouldn't be too bad, but the complex types would require a new serialization format. I don't really want to encourage Converter implementations that only have partial support for the API.
3. I was thinking we might want to also improve the file connector so it can demonstrate more complex data. For example, we could report a struct that included extra information like line number (maybe only added if a config flag is set).

ewencp committed:
Move and update outdated comment about handing of types for BYTES type in Copycat.

gwenshap commented:
1. Just copying log lines to Kafka is a valid use-case that can enjoy some simplicity. Note that we have String and Int serializers in o.a.k.common, so clearly we acknowledge that these are a thing.
2. I see your point about partial support. I thought we do want to support simple cases that don't require everything we provide.
3. Is it just for educational value? because I prefer the default connector example to be as simple as humanly possible.

gwenshap commented on copycat/api/src/main/java/org/apache/kafka/copycat/storage/Converter.java:
```diff
@@ -21,28 +21,37 @@
      * @return
      */
-    T fromCopycatData(Schema schema, Object value);
+    byte[] fromCopycatData(String topic, Schema schema, Object value);
```
wait what? why would we need a topic?

ewencp replied:
This is useful info to have during serialization and is included in the `Serializer` interface as well. Useful, e.g., for schema validation/registration.

ewencp committed:
Add StringConverter for using Copycat with raw strings.

ewencp commented:
@gwenshap Fair point about directly copying logs in as plain strings -- totally reasonable. However, I encountered an interesting problem -- both using Strings and when I turn on schema-less mode in the JsonConverter. Both will fail the system test if it's switched to use those Converters.

The reason is offsets. The offsets use the same key and value converters as regular data. Which means with a StringConverter, they will always be converted to Strings, which is not very helpful when you try to reload them...

One obvious choice is to add separate settings for offset serializer and deserializer. That may not be that bad, I had that previously and only removed it to reduce the number of required configs. However, it is also a bit odd to have to use multiple types of converters, and with some you'll just get random errors from connectors, e.g. this one from the test I mentioned:

```
org.apache.kafka.copycat.errors.CopycatException: Unexpected offset schema.
    at org.apache.kafka.copycat.file.FileStreamSourceTask.poll(FileStreamSourceTask.java:72)
    at org.apache.kafka.copycat.runtime.WorkerSourceTask$WorkerSourceTaskThread.execute(WorkerSourceTask.java:292)
    at org.apache.kafka.copycat.util.ShutdownableThread.run(ShutdownableThread.java:82)
```

This one is caught in a sane way because I explicitly verified the schema of the offset returned by the lookup in the FileStreamSourceTask:

```
                    if (!offsetWithSchema.schema().equals(OFFSET_VALUE_SCHEMA))
                        throw new CopycatException("Unexpected offset schema.");
                    Long lastRecordedOffset = (Long) offsetWithSchema.value();
```

But if I hadn't you would have just gotten a ClassCastException (which is the error that left a bunch of people confused about why serialization wasn't working when they were using the default serializer incorrectly and for which we added special error handling code to clarify in the new producer).

Not sure what the best solution is here. We could:
1. Add the `offset.key.converter` and `offset.value.converter` settings and not provide defaults. Users have to understand that they need to configure all of these, but that at least provides an opportunity to point out pitfalls since they'll need to read some docs about the settings and we can add notes in the sample configs.
2. Add the settings back and default to the `key.converter` and `value.converter` settings. This makes it easier in some cases, but error prone if they want to use non-schema-preserving converters. The user needs to understand the details of how capable different converters are or they are likely to end up with buggy configurations.
3. Do something with annotations/extra methods on converter classes so there is a way to check whether they have the necessary support. Annotations seem cleaner, but I don't know that it's sufficient because the JsonConverter can operate in both schema and schema-less modes. However, at least in schema-less mode it is better than StringConverter at preserving types -- connectors could potentially work around the missing schemas by assuming null schemas match and then catch any exceptions caused by mismatching types.

Another related issue is that if you need to configure any of these differently and they have settings with the same name, you are currently unable to because we just pass the worker configs directly to all the converters.

ewencp committed:
Restrict offset format.

Restrict offsets to be serialized via schemaless objects, but validate that they
are maps with string keys and primitive values. This is enforced at the Copycat
level rather than validating a Schema.

Additionally, parameterize copycat system test to work with and without schemas,
and to validate both JSON and String converters.

ewencp commented:
Just to keep readers of this thread in the loop, the approach we decided on offline is to separate out the configuration, make it default to using a format that we can guarantee serializes/deserializes with good fidelity, indicate that the setting is low importance and shouldn't be changed without great care. The offset commit API will change a bit. The framework will handle ensuring we maintain sufficient information about the writer schema for offsets that, given a list of schemas to use for reading, we are able to properly deserialize and return the data. This requires connectors to be careful about maintaining compatibility between versions, but this isn't new -- they always would have had to support multiple offset schemas if they change the format. The new solution just makes this more explicit in the APIs.

ewencp commented:
@gwenshap Updated to use the new approach to offsets we discussed. Also updated system test to be parameterized and test JSON and String serializers, including JSON with schemas and without schemas. Hopefully this covers enough bases that it's fairly good confirmation that we can make this work.

I'm honestly not enthusiastic about how the code for reading offsets looks since it's a bunch of checking for nulls and validating specific types, but maybe this is the best solution when we have to balance data with good schemas vs schemaless data.

I'm thinking maybe we can provide some utilities for converting maps with string fields to structs with some basic checks in place so it'll make connector implementors jobs' easier. What do you think about that?

ewencp committed:
Checkstyle fixes

gwenshap commented on config/copycat-standalone.properties:
```diff
@@ -16,12 +16,21 @@
 
+# The converters specify the format of data in Kafka and how to translate it into Copycat data. Every Copycat user will
+# need to configure these based on the format they want their data in when loaded from or stored into Kafka
 key.converter=org.apache.kafka.copycat.json.JsonConverter
```
Do we need the schemas.enable config for the data converter too?

ewencp replied:
It's right below this -- both `key.converter.schemas.enable=true` and `value.converter.schemas.enable=true`.

This is a new configuration convention which is something we might want to discuss. This was the best way I could come up with to avoid the potential conflicts.

gwenshap replied:
Right. I missed that. 

We can discuss, but I have to admit that I can't think of anything better either. 

We may want to document (for users) what this configuration mean / does.

gwenshap commented on copycat/api/src/main/java/org/apache/kafka/copycat/data/CopycatSchema.java:
```diff
@@ -23,20 +23,37 @@
+    /**
+     * Maps Schema.Types to a list of Java classes that can be used to represent them.
+     */
+    private static final Map<Type, List<Class>> SCHEMA_TYPE_CLASSES = new HashMap<>();
```
Maybe some explanation of when would a list with more than one item be useful?
The use-case is not completely obvious (You have schemas, they evolved, but you don't have a schema repository so you can't get the right version of the schema from an external system, IIRC).

gwenshap replied:
Sorry about prev. comment. got confused :)
This one is for bytes.

gwenshap commented on copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSourceTask.java:
```diff
@@ -66,14 +63,14 @@ public void start(Properties props) {
+                Map<String, Object> offset = context.offsetStorageReader().offset(Collections.singletonMap(FILENAME_FIELD, filename));
+                if (offset != null) {
+                    Object lastRecordedOffset = offset.get(POSITION_FIELD);
+                    if (lastRecordedOffset != null && !(lastRecordedOffset instanceof Long))
```
I hear ya on offset-reading requiring all sorts of boiler-plate that all connectors will need to suffer from.

gwenshap commented on copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSourceTask.java:
```diff
@@ -193,4 +190,12 @@ public void stop() {
         }
     }
+
+    private Map<String, String> offsetKey(String filename) {
```
Do you think we can turn these into more generic offset utilities that all connectors can share? 
I saw in your comments that you had some ideas.

Do you think it belongs in this patch? or a followup?

ewencp replied:
These are connector-specific, there's not a ton we can do to simplify/share this since `Collections.singletonMap` already made these methods quite concise. It's a good idea for connectors to have methods like these since they ensure all cases where an offset key or value is constructed are consistent.

I think the only thing we might be able to provide to help with this is something like `MapBuilder` so each connectors `offsetKey` and `offsetValue` method can be simplified by chaining, e.g. something like:

```
private Map<String, Object> offsetKey(String namespace, String filename) {
    return MapBuilder.builder().put(NAMESPACE_FIELD, namespace).put(FILENAME_FIELD, filename).build();
}
```

That's not at all critical though. It's just a very slightly nicer API, so I don't think it's all that important to include in this patch.

gwenshap replied:
Sounds good.

gwenshap commented:
LGTM :+1:
